File: Math/2. Add Two Numbers/Add Two Numbers.cpp
/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode(int x) : val(x), next(NULL) {}
* };
*/
class Solution {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {

		ListNode* result = new ListNode(-1);
		ListNode* aux = result; //辅助，用来处理中间链表
		int carry = 0;
		while (l1 != nullptr || l2 != nullptr || carry != 0){

			int a = l1 ? l1->val : 0; //判断链表当前是否为空，如果为空，则值为0
			int b = l2 ? l2->val : 0;
			int sum = a + b + carry;
			carry = sum / 10; //进位用取模来获得
			aux->next = new ListNode(sum % 10); //通过取模消除进位的影响
			aux = aux->next;
			if (l1)
				l1 = l1->next;
			if (l2)
				l2 = l2->next;
		}

		return result->next;
	}
};

File: Math/2. Add Two Numbers/Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(-1)
        aux = result
        carry = 0
        while (l1 or l2 or carry):
            a = l1 and l1.val or 0
            b = l2 and l2.val or 0
            sum = a + b + carry
            carry = sum / 10
            aux.next = ListNode(sum % 10)
            aux = aux.next
            if (l1):
                l1 = l1.next
            if (l2):
                l2 = l2.next

        return result.next
